src/main/java/org/jenkinsci/plugins/workflow/steps/ErrorInfoStep.java

```diff
@@ -0,0 +1,193 @@
+import org.jenkinsci.plugins.workflow.actions.ErrorAction;
+import org.jenkinsci.plugins.workflow.actions.LogAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graph.FlowScanner;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+/**
+ * Step to supply contextual information about an error that has been caught.
+ */
+public class ErrorInfoStep extends AbstractStepImpl {
+
+    public final Throwable error;
+
+    @DataBoundConstructor public ErrorInfoStep(Throwable error) {
+        this.error = error;
+    }
+
+    public static class Execution extends AbstractSynchronousStepExecution<ErrorInfo> {
+
+        private static final long serialVersionUID = 1;
+        @Inject private transient ErrorInfoStep step;
+        @StepContextParameter private transient FlowExecution execution;
+
+        @Override protected ErrorInfo run() throws Exception {
+            return new ErrorInfo(step.error, execution);
+        }
+
+    }
+
+    @Extension public static class DescriptorImpl extends AbstractStepDescriptorImpl {
+
+        public DescriptorImpl() {
+            super(Execution.class);
+        }
+
+        @Override public String getFunctionName() {
+            return "errorInfo";
+        }
+
+        @Override public String getDisplayName() {
+            return "Calculate information about an error";
+        }
+
+        // TODO blank config.jelly
+
+    }
+
+    public static class ErrorInfo implements Serializable {
+
+        private static final long serialVersionUID = 1;
+        private final Throwable error;
+        private transient FlowExecution execution;
+        private final FlowExecutionOwner executionOwner;
+
+        ErrorInfo(Throwable error, FlowExecution execution) {
+            this.error = error;
+            this.execution = execution;
+            executionOwner = execution.getOwner();
+        }
+
+        private FlowExecution getExecution() throws IOException {
+            if (execution == null) {
+                execution = executionOwner.get();
+            }
+            return execution;
+        }
+
+        /**
+         * Finds a node which threw this exception or one of its causes.
+         * Note that {@link Throwable#equals} is just pointer equality,
+         * which we cannot use since we may be loading deserialized exceptions,
+         * so we compare by stack trace instead.
+         */
+        private @CheckForNull FlowNode getNode() throws IOException {
+            final Set<String> stackTraces = new HashSet<>();
+            for (Throwable t = error; t != null; t = t.getCause()) {
+                stackTraces.add(Functions.printThrowable(t));
+            }
+            Predicate<FlowNode> threwException = new Predicate<FlowNode>() {
```

Comment:
I would inline this variable. Just make sure it is safe for lambda conversion when we switch to `-source 8`. Most likely this is satisfied by just verifying that there are no method overloads which vary only in the declared type of the SAM they accept. (This is where we screwed up with `ACL.impersonate`.)